### PR TITLE
Add Linux code paths for PowerShell install/uninstall scripts

### PR DESCRIPTION
**Pull request by danieljurek**

### CODE REVIEW

File: cli/installer/install-azd.ps1 [hunk 1]
```diff
@@ -118,39 +190,49 @@ Remove-Item $tempFolder -Recurse -Force | Out-Null
...
-            # SetEnvironmentVariable broadcasts the "Environment" change to
-            # Windows and is NOT destructive (e.g. expanding variables)
-            [Environment]::SetEnvironmentVariable(
+if (!$NoPath -and !(isLinuxOrMac)) {
+    # Wrap the Microsoft.Win32.Registry calls in a script block to prevent the
+    # type intializer from attempting to initialize those objects in non-Windows
+    # environments.
+    . {
+        try {
+            $registryKey = [Microsoft.Win32.Registry]::CurrentUser.OpenSubKey('Environment', $false)
```
Comment (danieljurek) [inline cli/installer/install-azd.ps1:199]: Try wrapping just the offended calls in script blocks. Curious if that also mitigates the issue.

Comment (danieljurek) [reply cli/installer/install-azd.ps1:199]: Wrapping just the `Microsoft.Win32` namespace calls works.

File: cli/installer/install-azd.ps1 [hunk 2]
```diff
@@ -76,33 +113,68 @@ if ($DryRun) {
...
+    $outputFilename = "$InstallFolder/azd"
+}
+
 try {
-    Copy-Item "$tempFolder/unzip/azd-windows-amd64.exe" "$InstallFolder/azd.exe" | Out-Null
+    if (isLinuxOrMac) {
+        test -w "$InstallFolder/"
+        if ($LASTEXITCODE) {
+            Write-Host "Writing to $InstallFolder/ requires elevated permission. You may be prompted to enter credentials."
+            sudo cp "$tempFolder/decompress/$binFilename" $outputFilename
```
Comment (weshaggard) [inline cli/installer/install-azd.ps1:167]: I assume cp here isn't an alias but instead the native cp command.

File: cli/installer/install-azd.ps1 [hunk 3]
```diff
@@ -76,33 +113,68 @@ if ($DryRun) {
...
+
+$outputFilename = "$InstallFolder/azd.exe"
+if (isLinuxOrMac) {
+    $outputFilename = "$InstallFolder/azd"
+}
+
 try {
-    Copy-Item "$tempFolder/unzip/azd-windows-amd64.exe" "$InstallFolder/azd.exe" | Out-Null
+    if (isLinuxOrMac) {
+        test -w "$InstallFolder/"
```
Comment (weshaggard) [inline cli/installer/install-azd.ps1:164]: What is this test command?

File: cli/installer/uninstall-azd.ps1 [hunk 4]
```diff
@@ -1,52 +1,87 @@
...
-            'PATH', `
-            $newPath, `
-            [EnvironmentVariableTarget]::User `
-        )
+if (isLinuxOrMac) {
+    $installLocation = "$InstallFolder/azd"
+
+    if (!(Test-Path $installLocation)) {
+        Write-Host "azd is not installed at $installLocation. To install run:"
+        Write-Host "pwsh -c `"Invoke-RestMethod 'https://aka.ms/install-azd.ps1' | Invoke-Expression`"'"
```
Comment (weshaggard) [inline cli/installer/uninstall-azd.ps1:22]:
```suggestion
        Write-Host "pwsh -c `"Invoke-RestMethod 'https://aka.ms/install-azd.ps1' | Invoke-Expression`""
```

File: eng/pipelines/release-cli.yml [hunk 5]
```diff
@@ -479,62 +497,68 @@ stages:
...
         strategy:
           matrix:
-            Windows:
-              Pool: azsdk-pool-mms-win-2019-general
-              OSVmImage: MMS2019
-              TestShell: cmd
-              TestCommand: >
-                cmd /c test-windows-install.cmd %BASEURL%
-            Linux:
+            LinuxSh:
```
Comment (weshaggard) [inline eng/pipelines/release-cli.yml:500]: Is Sh the right name given you are testing the ps1 script here?

Comment (danieljurek) [reply eng/pipelines/release-cli.yml:500]: Also of note: Investigate centos support matrix.

**Review by weshaggard — APPROVED** [review]
Couple comments but otherwise looks good.

File: cli/installer/install-azd.ps1 (hunk 3 above)
Comment (ellismg) [reply cli/installer/install-azd.ps1:164]: It's part of the POSIX standard (https://www.mankier.com/1p/test) and can be used to evaluate conditional expressions.

File: cli/installer/install-azd.ps1 [hunk 6]
```diff
@@ -55,17 +56,53 @@ param(
...
-    Write-Error "This bootstrap script is intended to run on Windows. Use the install-azd.sh instead."
-    exit 1
+function isLinuxOrMac {
+    return $IsLinux -or $IsMacOS
+}
+
+if (!$InstallFolder) {
+    $InstallFolder = "$($env:LocalAppData)\Programs\Azure Dev CLI"
+    if (isLinuxOrMac) {
+        $InstallFolder = "/usr/local/bin"
```
Comment (ellismg) [inline cli/installer/install-azd.ps1:71]: Nothing blocking but longer term I wonder if this should be in some part of the user's HomeDir, as to not require `sudo`.  If we do that we probably want to read https://wiki.archlinux.org/title/XDG_Base_Directory and see what location we should default to (I would guess `XDG_DATA_HOME`).

**Review by ellismg — APPROVED** [review]

File: cli/installer/install-azd.ps1 (hunk 6 above)
Comment (danieljurek) [reply cli/installer/install-azd.ps1:71]: https://github.com/Azure/azure-dev/issues/219

File: cli/installer/install-azd.ps1 (hunk 2 above)
Comment (danieljurek) [reply cli/installer/install-azd.ps1:167]: It is the native `cp` command (`/usr/bin/cp` in Ubuntu 20.04). If `sudo` and/or `cp` are overridden in pwsh then behavior here will not be predictable.

File: cli/installer/install-azd.ps1 (hunk 3 above)
Comment (danieljurek) [reply cli/installer/install-azd.ps1:164]: This is checking for writability of the install folder using the `test` command. This check is done to determine whether to elevate privileges. We could switch this around and attempt to write, and attempt to handle the error by elevating privileges and trying again.

**Review by wbreza — APPROVED** [review]
LGTM

**Pull request merged**